distributed/tests/test_worker.py

```diff
@@ -1634,3 +1634,27 @@ def bad_heartbeat_worker(*args, **kwargs):
                 else:
                     assert w.status == "closed"
     assert "Heartbeat to scheduler failed" in logger.getvalue()
+
+
+@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)])
+async def test_delete_data(client, s, w):
+    df = dask.datasets.timeseries(dtypes={"x": int, "y": float}, freq="1s").reset_index(
+        drop=True
+    )
+
+    await client.compute(df.map_partitions(lambda df: df.__sizeof__()))
+
+    df2 = df.persist()
+    await df2
+
+    def _check_data(empty):
+        if empty is True:
+            assert len(get_worker().data) == 0
+        else:
+            assert len(get_worker().data) > 0
+
+    del df
+    await client.run(_check_data, False)
+
+    del df2
+    await client.run(_check_data, True)
```

Comment:
To simplify this logic I recommend that you just check `w.data`

Reply:
I assume you're suggesting `assert len(w.data) > 0`/`assert len(w.data) == 0`, is that right? If so, that unfortunately fails on the second assertion, could this be some async black magic? E.g., `client.run` will force some syncing while `w.data` is still outdated?

Reply:
client.run doesn't force any synchronization, but maybe the GC hasn't run yet at this point.  Regardless we should probably make this robust to GC taking a while (GC delays are a major source of intermittent testing failures for us).  Most tests of this form look like the following:

```python
start = time()
while w.data:
    await asyncio.sleep(0.01)
    assert time() < start + 2
```

Reply:
Thanks for the explanation @mrocklin , your suggestion works and that's fixed in latest commits.